src/Components/InfiniteScroll.tsx

```diff
@@ -55,37 +69,140 @@ const DisplayPreviousPage = compose<IDisplayPreviousPageProps<any>, IDisplayPrev
   }),
 );
 
+interface IContentPlaceHolderProps {
+  height: number;
+}
+
+const DefaultContentPlaceHolder = ({ height }: IContentPlaceHolderProps) => {
+  const style = {
+    height,
+    borderBottom: "solid 1px #999",
+  };
+
+  return (
+    <div style={style} />
+  );
+};
+
 /**
- * Adds infinite scroll capability to a container and model component, querying for the next page of results when scrolling down.
+ * Adds infinite and virtual scroll capability to a container and model component,
+ * querying for the next page of results when scrolling down (or up if virtual scrolling).
  * The container component must be scrollable.
  */
 export const InfiniteScroll = compose<IInfiniteScrollInternalProps<any>, IInfiniteScrollProps<any>>(
   setDisplayName("InfiniteScroll"),
   defaultProps({
     debounceTime: 200,
+    disableVirtualScrolling: false,
+    contentPlaceHolderComponent: DefaultContentPlaceHolder,
   }),
-  withState<IInfiniteScrollProps<any>, IQueryBuilder, "query", "updateQuery">("query", "updateQuery", ({ query, modelName }) => (
-    query instanceof QueryBuilder
+  withProps(
+    {
+      currentPageHeightRef: React.createRef(),
+      previousPageHeightRef: React.createRef(),
+      nextPageHeightRef: React.createRef(),
+    }),
+  withStateHandlers<IInfiniteScrollHeightMapProps, { recordPageHeight }, IInfiniteScrollInternalProps<any>>(
+    { pageHeightMap: {}, estimatedPageHeight: 0 },
+    {
+      recordPageHeight: () =>
+        (pageHeightMap: { [key: string]: number }) => ({
+          pageHeightMap,
+          estimatedPageHeight: average(Object.values(pageHeightMap)),
+        }),
+    },
+  ),
+  withState<IInfiniteScrollProps<any>, IQueryBuilder, "query", "updateQuery">("query", "updateQuery", ({ query, modelName }) => {
+    return query instanceof QueryBuilder
       ? query
-      : new QueryBuilder(modelName, query as IQueryParams)
-  )),
+      : new QueryBuilder(modelName, query as IQueryParams);
+  }),
   withModelQuery(),
+  withPropsOnChange(["query", "estimatedPageHeight"], ({ query, pageHeightMap, estimatedPageHeight }) => {
+    let previousPlaceHolderHeight = 0;
+    let nextPlaceHolderHeight = 0;
+
+    for (let x = query.response.currentPage - 2; x > 0; x--) {
+      previousPlaceHolderHeight += x in pageHeightMap
+        ? pageHeightMap[x]
+        : estimatedPageHeight;
+    }
+
+    for (let x = query.response.currentPage + 2; x <= query.response.totalPages; x++) {
+      nextPlaceHolderHeight += x in pageHeightMap
+        ? pageHeightMap[x]
+        : estimatedPageHeight;
+    }
+
+    return {
+      previousPlaceHolderHeight,
+      nextPlaceHolderHeight,
+    };
+  }),
+  withProps(({
```

Comment:
I think we should use `withPropsOnChange` so we do not redeclare `updatePageHeightMap` whenever any of the props change.

Reply:
I'm actually wondering if this should be withHandlers instead

Reply:
changed to withHandlers